ocaml/xapi/quicktest_cbt.ml

```diff
@@ -0,0 +1,177 @@
+        VDI.data_destroy ~session_id ~rpc:!rpc ~self:newvdi;
+        debug data_destroy_test "Destroy snapshot VDI data";
+        test_compare ~test
+          (VDI.get_type ~session_id ~rpc:!rpc ~self:newvdi)
+          `cbt_metadata
+          ~msg:"VDI.data_destroy failed to update VDI.type";
+        let content_id_str = "/No content: this is a cbt_metadata VDI/" in
+        test_compare ~test
+          (VDI.get_other_config ~session_id ~rpc:!rpc ~self:newvdi |> List.assoc "content_id")
+          content_id_str
+          ~msg:"VDI.data_destroy failed to update VDI.content_id";
+        VDI.destroy ~session_id ~rpc:!rpc ~self:newvdi;
+        debug data_destroy_test "Snapshot destroyed.";
+        success data_destroy_test
+      with e -> report_failure e "data_destroy" data_destroy_test in
+
+    (* Check VDI.(copy, clone, and snapshot) all properly update cbt_enabled *)
+    let vdi_clone_copy_snapshot_test ~sR ~vDI =
+      let vdi_duplicate_test = make_test "Testing VDI.clone,copy,snapshot" 4 in
+      try
+        start vdi_duplicate_test;
+        let test = vdi_duplicate_test in
+        VDI.enable_cbt ~session_id ~rpc:!rpc ~self:vDI;
+        test_assert ~test
+          (get_cbt_status vDI)
+          ~msg:"VDI.enable_cbt failed";
+
+        let vdi_clone = VDI.clone ~session_id ~rpc:!rpc ~vdi:vDI ~driver_params:[] in
+        test_assert ~test
+          (get_cbt_status vdi_clone)
+          ~msg:"VDI.clone failed to update cbt_enabled";
+
+        let vdi_snapshot = VDI.snapshot ~session_id ~rpc:!rpc ~vdi:vDI ~driver_params:[] in
+        test_assert ~test
+          (get_cbt_status vdi_snapshot)
+          ~msg:"VDI.snapshot failed to update cbt_enabled";
+
+        (* Test VDI.copy for copying from existing to fresh VDI in same SR *)
+        let vdi_copy_fresh = VDI.copy ~session_id ~rpc:!rpc ~vdi:vDI ~base_vdi:(Ref.null) ~into_vdi:(Ref.null) ~sr:sR in
+        test_assert ~test
+          (get_cbt_status vdi_copy_fresh)
+          ~msg:"VDI.copy failed to update cbt_enabled";
+
+        (* Test VDI.copy for copying from one existing to another *)
+        let into_vdi = make_vdi_from sR in
+        let vdi_copy_existing = VDI.copy ~session_id ~rpc:!rpc ~vdi:vDI ~base_vdi:(Ref.null) ~into_vdi ~sr:(Ref.null) in
+        test_assert ~test
+          (get_cbt_status vdi_copy_existing)
+          ~msg:"VDI.copy failed to update cbt_enabled";
+        (* Test VDI copied into? *)
+        test_assert ~test
+          (get_cbt_status into_vdi)
+          ~msg:"VDI.copy failed to update cbt_enabled";
+
+        List.iter (fun self -> VDI.destroy ~session_id ~rpc:!rpc ~self)
+          [ vdi_clone ; vdi_snapshot ; vdi_copy_fresh ; vdi_copy_existing ; into_vdi];
+        success vdi_duplicate_test
+      with e -> report_failure e "VDI_copy_clone_snapshot" vdi_duplicate_test in
+
+
+    let run_test_suite ~sR ~vDI =
+      let sr_ops = (SR.get_allowed_operations ~session_id ~rpc:!rpc ~self:sR) in
+      List.iter
```

Comment:
Just a suggestion:

```
[ (fun () -> enable_disable_cbt_test ~vDI), [ `vdi_enable_cbt ; `vdi_disable_cbt ]
; (fun () -> vdi_clone_copy_snapshot_test ~sR ~vDI), [ `vdi_enable_cbt ; `vdi_create ; `vdi_clone ; `vdi_snapshot ; `vdi_destroy ]
; (fun () -> vdi_data_destroy_test ~vDI), [ `vdi_enable_cbt ; `vdi_data_destroy ; `vdi_snapshot ; `vdi_destroy ]
]
|> List.iter (fun (test, ops) -> ... )
```

This way data flows from top to bottom.

Reply:
I'd go the next step and do 
```ocaml
[ (fun () -> enable_disable_cbt_test ~vDI),
  [ `vdi_enable_cbt ; `vdi_disable_cbt ]
; (fun () -> vdi_clone_copy_snapshot_test ~sR ~vDI),
  [ `vdi_enable_cbt ; `vdi_create ; `vdi_clone ; `vdi_snapshot ; `vdi_destroy ]
; (fun () -> vdi_data_destroy_test ~vDI),
  [ `vdi_enable_cbt ; `vdi_data_destroy ; `vdi_snapshot ; `vdi_destroy ]
]
|> List.iter (fun (test, ops) -> ... )